src/course-outline/CourseOutline.jsx

```diff
@@ -158,26 +159,19 @@ const CourseOutline = ({ courseId }) => {
     });
   };
 
-  const unitsIdPattern = useMemo(() => {
-    let pattern = '';
-    sections.forEach((section) => {
-      section.childInfo.children.forEach((subsection) => {
-        subsection.childInfo.children.forEach((unit) => {
-          if (pattern !== '') {
-            pattern += `,${unit.id}`;
-          } else {
-            pattern += unit.id;
-          }
-        });
-      });
-    });
-    return pattern;
-  }, [sections]);
-
-  const {
-    data: unitsTagCounts,
-    isSuccess: isUnitsTagCountsLoaded,
-  } = useUnitTagsCount(unitsIdPattern);
+  // ToDo: Discuss in the PR
+  // We could use this to prefetch the tags count for all units/blocks in the course
+  // But to prevent refetching the same data when the component is re-rendered, we should
+  // configure the query stale time to a reasonable value.
+  // IMHO, we should just let each card to fetch the tags count by itself. This is simpler and
+  // is working fine.
+  // const contentTagPatternMutation = useContentPatternTagsCount();
+  //
+  // useEffect(() => {
+  //   if (courseId) {
+  //     contentTagPatternMutation.mutate(`${courseId.replace('course-v1', 'block-v1', 1)}*`);
+  //   }
+  // }, [courseId]);
```

Comment:
@ChrisChV I think we don't need this. The multiple calls were a problem in the Studio, but I think it works smoothly here. What do you think? CC @bradenmacdonald

Reply:
@rpenido I think we don't need it for now, it's a lot for an MVP. Unless a performance analysis says flatly otherwise :smile: 

About let each cart to fetch the tag count, I think it would increase the number of calls made to the API, one for each card. Using one or the other approach I don't see much difference, but the increase in the number of calls is what makes me lean toward just making one call for all

CC @bradenmacdonald

Reply:
I think it is working smoothly even when making A LOT of API calls.

[tag_course_outline.webm](https://github.com/open-craft/frontend-app-course-authoring/assets/849463/629c4802-2191-4ad9-a7b1-e1c52406502a)

We could tweak it by setting a stale time for each query, but I think we should do this for the whole app instead of on a query-by-query basis.

Reply:
Our API to get the counts for every item in the course should be pretty fast. I would lean toward just using that everywhere, so there's only one API request. Fetching hundreds of separate API requests on the outline page may work fine on devstack but I suspect will cause issues on production. Especially if people don't have HTTP/2 working.

The only downside is when you invalidate the tags of a particular item, you also have to invalidate the counts of everything in the whole course and fetch it again, but I think it's fine.

Also to be clear, each component (e.g. each unit in the outline) should use the same API hook to "fetch tag counts for the whole course". React Query will de-duplicate this to only a single request. Then take the result, search for the ID of the current unit, and display that count, if any.

Reply:
> The only downside is when you invalidate the tags of a particular item, you also have to invalidate the counts of everything in the whole course and fetch it again, but I think it's fine.

It is possible to solve this: A mutation prefetch the data and populatse the queries via `setQueryData`. So we get all counts on first load and then each component have their own query after that.

https://github.com/open-craft/frontend-app-course-authoring/pull/30/files#diff-0fb42c9f642840f13f2018e2ab11fc2386e22caa9baa226843197bf16d4151cfR10-R24

The problem with this code is that mounting the component (the unit cards when expanding) trigger the refetch for the query. This could be fixed setting a proper `staleTime`, but it will need more time.

I will leave it for now and use a single useQuery as you both suggested (but we could go back to this mixed approach if someday the single query hit us).

Thank you for the feedback @ChrisChV @bradenmacdonald !

Reply:
@rpenido I thought about that too, and it's what I wanted to suggest, but I think there's a problem with it. During the pre-fetch, if the individual cards render before the prefetch is complete, they will still each individually make a request to the API. I looked a lot of the React Query docs like https://tanstack.com/query/v4/docs/framework/react/guides/prefetching and https://tkdodo.eu/blog/seeding-the-query-cache but I couldn't figure out a way to tell React Query _not_ to load the per-item query keys while the prefetch is happening, but still to load them later if they get invalidated.

Reply:
> but I couldn't figure out a way to tell React Query _not_ to load the per-item query keys while the prefetch is happening, but still to load them later if they get invalidated.

We could call the prefetch mutation at the CourseOutline page and render (the page or the course tree) only after we get the mutation response, as we do with all course data (sections, subsections, etc.).

I think it will be possible if we spend some time with it, but I don't think we should do it now. The overall responsiveness is pretty good. I don't want to add complexity to solve a problem that we don't have yet. :slightly_smiling_face: